executor/pipelined_window.go

```diff
@@ -0,0 +1,462 @@
+	for i, windowFunc := range p.windowFuncs {
+		if slidingWindowAggFunc, ok := windowFunc.(aggfuncs.SlidingWindowAggFunc); ok {
+			p.slidingWindowFuncs[i] = slidingWindowAggFunc
+		}
+	}
+}
+
+// slide consumes more rows, and it returns the number of more rows needed in order to proceed, more = -1 means it can only
+// process when the whole partition is consumed, more = 0 means it is ready now, more = n (n > 0), means it need at least n
+// more rows to process.
+func (p *processor) consume(ctx sessionctx.Context, rows []chunk.Row) (more int64, err error) {
+	num := uint64(len(rows))
+	p.rowCnt += num
+	p.rows = append(p.rows, rows...)
+	rows = rows[:0]
+	if p.end.UnBounded {
+		if !p.whole {
+			// we can't proceed until the whole partition is consumed
+			return -1, nil
+		}
+	}
+	// let's just try produce
+	return 0, nil
+}
+
+// finish is called upon a whole partition is consumed
+func (p *processor) finish() {
+	p.whole = true
+}
+
+func (p *processor) getStart(ctx sessionctx.Context) (uint64, error) {
+	if p.start.UnBounded {
+		return 0, nil
+	}
+	if p.isRangeFrame {
+		var start uint64
+		for start = p.lastStartRow; start < p.rowCnt; start++ {
+			var res int64
+			var err error
+			for i := range p.orderByCols {
+				res, _, err = p.start.CmpFuncs[i](ctx, p.orderByCols[i], p.start.CalcFuncs[i], p.getRow(start), p.getRow(p.curRowIdx))
+				if err != nil {
+					return 0, err
+				}
+				if res != 0 {
+					break
+				}
+			}
+			// For asc, break when the calculated result is greater than the current value.
+			// For desc, break when the calculated result is less than the current value.
+			if res != p.expectedCmpResult {
+				break
+			}
+		}
+		return start, nil
+	}
+	switch p.start.Type {
+	case ast.Preceding:
+		if p.curRowIdx > p.start.Num {
+			return p.curRowIdx - p.start.Num, nil
+		}
+		return 0, nil
+	case ast.Following:
+		return p.curRowIdx + p.start.Num, nil
+	default: // ast.CurrentRow
+		return p.curRowIdx, nil
+	}
+}
+
+func (p *processor) getEnd(ctx sessionctx.Context) (uint64, error) {
+	if p.end.UnBounded {
+		if p.whole {
+			return p.rowCnt, nil
+		}
+		return p.rowCnt + 1, nil
+	}
+	if p.isRangeFrame {
+		var end uint64
+		for end = p.lastEndRow; end < p.rowCnt; end++ {
+			var res int64
+			var err error
+			for i := range p.orderByCols {
+				res, _, err = p.end.CmpFuncs[i](ctx, p.end.CalcFuncs[i], p.orderByCols[i], p.getRow(p.curRowIdx), p.getRow(end))
+				if err != nil {
+					return 0, err
+				}
+				if res != 0 {
+					break
+				}
+			}
+			// For asc, break when the calculated result is greater than the current value.
+			// For desc, break when the calculated result is less than the current value.
+			if res == p.expectedCmpResult {
+				break
+			}
+		}
+		return end, nil
+	}
+	switch p.end.Type {
+	case ast.Preceding:
+		if p.curRowIdx >= p.end.Num {
+			return p.curRowIdx - p.end.Num + 1, nil
+		}
+		return 0, nil
+	case ast.Following:
+		return p.curRowIdx + p.end.Num + 1, nil
```

Comment:
Should we check the `end` whether it exceeded `p.whole`.

Reply:
The check is done in produce